File: 200-number-of-islands/number-of-islands.py
class Solution:
    def dfs(self, grid, vis, row, col):
        vis[row][col] = True
        direction = [(0, 1), (0, -1), (1, 0), (-1, 0)]
        for dr, dc in direction:
            nr, nc = dr+row, dc+col
            if 0<=nr<len(grid) and 0<=nc<len(grid[0]):
                if grid[nr][nc]=="1" and not vis[nr][nc]:
                    self.dfs(grid, vis, nr, nc)


    def numIslands(self, grid: List[List[str]]) -> int:
        n = len(grid)
        m = len(grid[0])
        visited = [[False] * m for _ in range(n)]
        count = 0
        for i in range(n):
            for j in range(m):
                if grid[i][j]=="1" and not visited[i][j]:
                    self.dfs(grid, visited, i, j)
                    count+=1
        return count

File: 200-number-of-islands/number-of-islands.cpp
class Solution {
public:
    void dfs(vector<vector<bool>> &vis, int row, int col, vector<vector<char>>& grid){
        vis[row][col] = true;
        vector<vector<int>> directions = {{0, 1}, {1, 0}, {0, -1}, {-1, 0}};
        for(int i=0; i<4; i++){
            int nrow = row+directions[i][0];
            int ncol = col+directions[i][1];
            if(nrow<grid.size() && ncol<grid[0].size()){
                if(grid[nrow][ncol]=='1' && !vis[nrow][ncol]) dfs(vis, nrow, ncol, grid);
            }                
        }
    }

    int numIslands(vector<vector<char>>& grid) {
        int row = grid.size();
        int col = grid[0].size();

        vector<vector<bool>> vis(row, vector<bool>(col, false));
        int count = 0;
        for(int i=0; i<row; i++){
            for(int j=0; j<col; j++){
                if(grid[i][j]=='1' && !vis[i][j]){
                    dfs(vis, i, j, grid);
                    count++;
                }
            }
        }

        return count;
    }
};